sdk/core/System.ClientModel/src/Internal/SSE/ServerSentEventReader.cs

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Internal;
+
+/// <summary>
+/// An SSE event reader that reads lines from an SSE stream and composes them
+/// into SSE events.
+/// See SSE specification: https://html.spec.whatwg.org/multipage/server-sent-events.html
+/// </summary>
+internal sealed class ServerSentEventReader
+{
+    private readonly Stream? _stream;
+    private readonly StreamReader? _reader;
+
+    public ServerSentEventReader(Stream stream)
+    {
+        // Creator of the reader has responsibility for disposing the stream
+        // passed to the reader constructor.
+
+        _stream = stream;
+        _reader = new StreamReader(stream);
+    }
+
+    /// <summary>
+    /// Synchronously retrieves the next server-sent event from the underlying stream, blocking until a new event is
+    /// available and returning null once no further data is present on the stream.
+    /// </summary>
+    /// <param name="cancellationToken"> An optional cancellation token that can abort subsequent reads. </param>
+    /// <returns>
+    ///     The next <see cref="ServerSentEvent"/> in the stream, or null once no more data can be read from the stream.
+    /// </returns>
+    public ServerSentEvent? TryGetNextEvent(CancellationToken cancellationToken = default)
+    {
+        if (_reader is null)
+        {
+            throw new ObjectDisposedException(nameof(ServerSentEventReader));
```

Comment:
Since the type is both internal and not disposable, could this exception be confusing?   It may be worth a comment to explain why we chose it.

Reply:
Ah, good catch.  We had this IDisposable before - I took that out, but I didn't take out all the related LOC.